### ISSUE TITLE
Revisit random seed options in ANTs

### ISSUE THREAD

**Issue by cookpa** [ANTsX/ANTs#1464]
Following discussion in #1462

I believe reproducibility can be achieved by setting the env var `ANTS_RANDOM_SEED` and using single-threaded computation to avoid numerical issues in the mutual information metric.

But the usage of some scripts might be misleading, eg in `antsBrainExtraction.sh`, "-u 0" fixes the Atropos random seed, but doesn't set a random seed to be used by all the other processes.

**Comment by cookpa** [comment]
Hi @ntustison , 

Thinking about this some more, I think `antsAI` wasn't the cause of the issue you were seeing recently, because it defaults to a fixed seed:

https://github.com/ANTsX/ANTs/blob/74afcc56cfdc98ffbb3d55b2487fa33e512fe8a5/Examples/antsAI.cxx#L1245-L1265

I think it was `antsRegistration` being variable. I'm thinking for now that we could set a fixed seed inside the script if "-u 0" is detected.

Longer term, it would be good to make all the programs consistent in how they handle random number generation. `antsRegistration` does, in decreasing order of priority,

1. Command line --random-seed <some seed> uses <some seed>
2. $ANTS_RANDOM_SEED
3. System time

Having this be consistent for all tools would be nice because it removes the need to add a random option explicitly to every executable call. I guess one criticism is that you could run a very complex pipeline with the same random numbers over and over, but I don't know if anyone cares to use ANTS_RANDOM_SEED outside of regression testing anyway.

**Comment by cookpa** [comment]
Solution in #1468 aims to maximize backwards compatibility and hopefully won't break anyone's regression tests

**Comment by TheChymera** [comment]
What's the situation with recent-ish ANTs (`2.4.3`)? is setting a global `ANTS_RANDOM_SEED` value analogous to using `--random-seed` for each `antsRegistration` call?

**Comment by cookpa** [comment]
@TheChymera yes, `ANTS_RANDOM_SEED` should be used unless `--random-seed` is specified at run time.

**Comment by cookpa** [comment]
The last part of my to-do list for this issue was `antsLongitudinalCorticalThickness.sh`, but because it uses `ANTS`, it doesn't give identical results even if ANTS_RANDOM_SEED is set. I tried fixing the seed here

https://github.com/ANTsX/ANTs/blob/3fa6cfb8eddb303ca69ba9e3abfd4566dcd011c0/ImageRegistration/ANTS_affine_registration.h#L260-L261

but it was insufficient.

Since deterministic results aren't a documented feature of the longitudinal script, I'm inclined to leave it alone.

**Comment by cookpa** [comment]
I think this is as close to complete as it can get
